File: app/src/main/java/wangdaye/com/geometricweather/ui/decotarion/GridMarginsDecoration.java
package wangdaye.com.geometricweather.ui.decotarion;

import android.content.Context;
import android.graphics.Rect;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Px;
import androidx.recyclerview.widget.RecyclerView;

import wangdaye.com.geometricweather.R;

public class GridMarginsDecoration extends RecyclerView.ItemDecoration {

    private @Px float marginsVertical;
    private @Px float marginsHorizontal;

    public GridMarginsDecoration(Context context, RecyclerView recyclerView) {
        this(context.getResources().getDimensionPixelSize(R.dimen.little_margin), recyclerView);
    }

    public GridMarginsDecoration(@Px float margins, RecyclerView recyclerView) {
        this(margins, margins, recyclerView);
    }

    public GridMarginsDecoration(@Px float marginsVertical, @Px float marginsHorizontal, RecyclerView recyclerView) {
        this.marginsVertical = marginsVertical;
        this.marginsHorizontal = marginsHorizontal;
        recyclerView.setClipToPadding(false);
        recyclerView.setPadding(
                (int) marginsHorizontal / 2,
                (int) marginsVertical / 2,
                (int) marginsHorizontal / 2,
                (int) marginsVertical / 2
        );
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view,
                               @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        super.getItemOffsets(outRect, view, parent, state);
        outRect.set(
                (int) (marginsHorizontal / 2),
                (int) (marginsVertical / 2),
                (int) (marginsHorizontal / 2),
                (int) (marginsVertical / 2)
        );
    }
}

File: app/src/main/java/org/breezyweather/common/ui/decorations/GridMarginsDecoration.kt
package org.breezyweather.common.ui.decorations

import android.content.Context
import android.graphics.Rect
import android.view.View
import androidx.annotation.Px
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import org.breezyweather.R

class GridMarginsDecoration(
    @field:Px @param:Px private val mMarginsVertical: Float,
    @field:Px @param:Px private val mMarginsHorizontal: Float,
    recyclerView: RecyclerView
) : ItemDecoration() {
    constructor(context: Context, recyclerView: RecyclerView) : this(context.resources.getDimensionPixelSize(R.dimen.little_margin).toFloat(), recyclerView)
    constructor(@Px margins: Float, recyclerView: RecyclerView) : this(margins, margins, recyclerView)

    init {
        recyclerView.clipToPadding = false
        recyclerView.setPadding(
            (mMarginsHorizontal / 2).toInt(),
            (mMarginsVertical / 2).toInt(),
            (mMarginsHorizontal / 2).toInt(),
            (mMarginsVertical / 2).toInt()
        )
    }

    override fun getItemOffsets(
        outRect: Rect, view: View,
        parent: RecyclerView, state: RecyclerView.State
    ) {
        super.getItemOffsets(outRect, view, parent, state)
        outRect.set(
            (mMarginsHorizontal / 2).toInt(),
            (mMarginsVertical / 2).toInt(),
            (mMarginsHorizontal / 2).toInt(),
            (mMarginsVertical / 2).toInt()
        )
    }
}